### PR TITLE
[R4ML-195] cache as.r4ml.matrix() by default

### PR DESCRIPTION
**Pull request by bdwyer2**
fixed issue with glm, lm, and potentially other algos when data is split but not cached and created test cases to verify this.

 
Developer's Certificate of Origin 1.1

       By making a contribution to this project, I certify that:

       (a) The contribution was created in whole or in part by me and I
           have the right to submit it under the Apache License 2.0; or

       (b) The contribution is based upon previous work that, to the best
           of my knowledge, is covered under an appropriate open source
           license and I have the right under that license to submit that
           work with modifications, whether created in whole or in part
           by me, under the same open source license (unless I am
           permitted to submit under a different license), as indicated
           in the file; or

       (c) The contribution was provided directly to me by some other
           person who certified (a), (b) or (c) and I have not modified
           it.

       (d) I understand and agree that this project and the contribution
           are public and that a record of the contribution (including all
           personal information I submit with it, including my sign-off) is
           maintained indefinitely and may be redistributed consistent with
           this project or the open source license(s) involved.

### CODE REVIEW

File: R4ML/tests/testthat/test-ml.model.glm.R [hunk 1]
```diff
@@ -115,3 +115,26 @@ test_that("r4ml.glm accuracy", {
                as.data.frame(r_predict_icpt)$r_predict, tol = .01)
 
 })
+
+test_that("r4ml.glm sampling check", {
+  # this is a regression test for R4ML-195
```
Comment (aloknsingh) [inline R4ML/tests/testthat/test-ml.model.glm.R]: May be we want to print it as the info msg about R4ML-195. spark does it too.

Comment (bdwyer2) [reply R4ML/tests/testthat/test-ml.model.glm.R]: What do you mean?

File: R4ML/tests/testthat/test-ml.model.glm.R [hunk 2]
```diff
@@ -115,3 +115,26 @@ test_that("r4ml.glm accuracy", {
...
+    omit.na = c("Distance", "ArrDelay"),
+    dummycodeAttrs = c("DayOfWeek")
+    )
+
+  df <- as.r4ml.matrix(df$data)
+
+  samples <- r4ml.sample(df, perc = c(0.7, 0.3))
+  train <- samples[[1]]
+
+  glm <- r4ml.glm(ArrDelay ~ ., data = train)
```
Comment (aloknsingh) [inline R4ML/tests/testthat/test-ml.model.glm.R]: we should test other algo mostly linear ones. so it should be move to another test file where there sampling is defined

File: R4ML/tests/testthat/test-ml.model.glm.R [hunk 3]
```diff
@@ -115,3 +115,26 @@ test_that("r4ml.glm accuracy", {
                as.data.frame(r_predict_icpt)$r_predict, tol = .01)
 
 })
+
+test_that("r4ml.glm sampling check", {
+  # this is a regression test for R4ML-195
+  df <- airline[1:500, c("Month", "DayofMonth", "DayOfWeek", "CRSDepTime",
```
Comment (aloknsingh) [inline R4ML/tests/testthat/test-ml.model.glm.R]: comments why 500 rows only?

File: R4ML/tests/testthat/test-ml.model.glm.R [hunk 4]
```diff
@@ -115,3 +115,26 @@ test_that("r4ml.glm accuracy", {
...
+    recodeAttrs = c("DayOfWeek"),
+    omit.na = c("Distance", "ArrDelay"),
+    dummycodeAttrs = c("DayOfWeek")
+    )
+
+  df <- as.r4ml.matrix(df$data)
+
+  samples <- r4ml.sample(df, perc = c(0.7, 0.3))
+  train <- samples[[1]]
+
```
Comment (aloknsingh) [inline R4ML/tests/testthat/test-ml.model.glm.R]: we want to give the best practices idea of using sampling before systemML algo call. so I think may be we should put as.r4ml.matrix here. But then it doesn't test the use case.

Comment (aloknsingh) [reply R4ML/tests/testthat/test-ml.model.glm.R]: may be we should test it also by putting before glm too

Comment (bdwyer2) [reply R4ML/tests/testthat/test-ml.model.glm.R]: `as.r4ml.matrix()` needs to go before we do sampling or we get this error:
```
> samples <- r4ml.sample(df, perc = c(0.7, 0.3))

Error: ERROR[r4ml.sample]: The specified dataset must either be a r4ml.frame, r4ml.matrix, or SparkDataFrame.
```

File: R4ML/tests/testthat/test-ml.model.glm.R (hunk 3 above)
Comment (bdwyer2) [reply R4ML/tests/testthat/test-ml.model.glm.R]: To save time by not running the entire dataset

File: R4ML/tests/testthat/test-r4ml.sampling.R [hunk 5]
```diff
@@ -79,3 +79,27 @@ test_that("r4ml.sampling k split\n", {
...
+    )
+
+  df <- as.r4ml.matrix(df$data)
+
+  samples <- r4ml.sample(df, perc = c(0.7, 0.3))
+  train <- samples[[1]]
+
+  r4_lm <- r4ml.lm(ArrDelay ~ ., data = train)
+  r4_glm <- r4ml.glm(ArrDelay ~ ., data = train)
+  r4_step_lm <- r4ml.step.lm(ArrDelay ~ ., data = train)
```
Comment (aloknsingh) [inline R4ML/tests/testthat/test-r4ml.sampling.R:104]: mlogit

File: R4ML/R/r4ml.sampling.R [hunk 6]
```diff
@@ -96,7 +96,7 @@ r4ml.sample <- function(data, perc, experimental=FALSE,
         r4ml.err(logSource, "Unsupported type " %++% data_type %++% " passed in")
       }
      
-      if (cache) {
+      if (cache & !casted_df@env$isCached) {
         dummy <- cache(casted_df)
```
Comment (aloknsingh) [inline R4ML/R/r4ml.sampling.R]: SparkR::cache

File: R4ML/R/r4ml.matrix.R [hunk 7]
```diff
@@ -73,24 +73,27 @@ setMethod("as.r4ml.matrix",
     } else {
       r4ml.err(logSource, "Use hydrar.ml.preprocess to convert r4ml.frame to numeric r4ml.matrix")
     }
-    result
+    if (cache & !result@env$isCached) {
+      result <- cache(result)
```
Comment (aloknsingh) [inline R4ML/R/r4ml.matrix.R]: SparkR::cache

File: R4ML/R/r4ml.matrix.R [hunk 8]
```diff
@@ -46,19 +46,19 @@ setClass("r4ml.matrix",
 #'
 #' Convert an object to a r4ml.matrix
 #' @param object A r4ml.frame, data.frame, or Spark DataFrame
+#' @param cache (logical) should the object be cached
 #' @return A r4ml.matrix
 #' @examples \dontrun{
-#' r4ml_frame <-  as.r4ml.frame(beaver1)
-#' r4ml_matrix <- as.r4ml.matrix(r4ml_frame)
+#' r4ml_matrix <- as.r4ml.matrix(datasets::beaver1)
```
Comment (aloknsingh) [inline R4ML/R/r4ml.matrix.R]: should we add example of cache? so that use know it?

File: R4ML/tests/testthat/test-r4ml.sampling.R (hunk 5 above)
Comment (bdwyer2) [reply R4ML/tests/testthat/test-r4ml.sampling.R:104]: The dataset I'm using isn't suited for mlogit

File: R4ML/R/ml.preprocess.R [hunk 9]
```diff
@@ -266,7 +266,9 @@ r4ml.ml.preprocess <- function(
     rhf = data
     if (is.omit.na) {
       rhf <- as.r4ml.frame(SparkR::dropna(data, cols = omit.na), repartition = FALSE)
-      ignore <- cache(rhf)
+      if (!rhf@env$isCached) {
+        ignore <- cache(rhf)
```
Comment (aloknsingh) [inline R4ML/R/ml.preprocess.R]: Can we pl change this to SparkR::cache

**Review by aloknsingh — APPROVED** [review]

**Pull request closed without merging**